payments-core/src/main/java/com/stripe/android/paymentsheet/PaymentSheetViewModel.kt

```diff
@@ -190,6 +182,47 @@ internal class PaymentSheetViewModel @Inject internal constructor(
                 stripeIntentValidator.requireValid(stripeIntent)
             }.fold(
                 onSuccess = {
+                    updatePaymentMethods(stripeIntent)
+                    resetViewState(stripeIntent, userErrorMessage = null)
+                },
+                onFailure = ::onFatal
+            )
+        }
+    }
+
+    /**
+     * Fetch the saved payment methods for the customer, if a [PaymentSheet.CustomerConfiguration]
+     * was provided.
+     * It will fetch only the payment method types accepted by the [stripeIntent] and defined in
+     * [SupportedSavedPaymentMethod].
+     */
+    @VisibleForTesting
+    fun updatePaymentMethods(stripeIntent: StripeIntent) {
+        viewModelScope.launch {
+            runCatching {
+                customerConfig?.let { customerConfig ->
+                    stripeIntent.paymentMethodTypes.mapNotNull {
+                        SupportedSavedPaymentMethod.fromCode(it)
+                    }.map {
+                        async {
+                            paymentMethodsRepository.get(
+                                customerConfig,
+                                it.type
+                            )
+                        }
+                    }.awaitAll().flatten().filter { paymentMethod ->
+                        SupportedSavedPaymentMethod.isSupported(paymentMethod).also { valid ->
```

Comment:
I might be reading this all wrong, but if you call paymentMethodRepository.get only for the SupportedSavePaymentMethods on line 205, then do you need to check if it is supported again?

Reply:
In line 205 we go through the types supported in the StripeIntent and filter out the types that payment sheet doesn't support. Here `SupportedSavedPaymentMethod.isSupported` looks into the content of `paymentMethod`. This check was done later when creating the `PaymentOptions`, I moved it here so it's done as early as possible and an error is logged.

Reply:
I am not sure I understand yet.  Are you suggesting that you could request one payment method type and then get a different payment method type back?  For instance request paymentMethods of type sofort, but get BLIK payment methods back?

Reply:
Just for future reference, if you request sofort you will get sofort back.